### ISSUE TITLE
option to customize attachShadow call?

### ISSUE THREAD

**Issue by amarajs** [hybridsjs/hybrids#13]
Hey guys,

I've run into some problems related to focusing distributed nodes. Specifically:

 - I want a `custom-button` element to delegate focus to the inner button instance when the custom button is focused.
 - I want a `custom-dialog` element with `<dialog><slot></slot></dialog>` to use the normal dialog focus logic (e.g. focusing the distributed node with the `autofocus` attribute, if one exists).

These behaviors [would occur automatically](https://w3c.github.io/webcomponents/spec/shadow/#focus) by passing `delegatesFocus` to the `attachShadow` call, but Hybrids' render method calls that, and I can't see an easy way to override that.

What are your thoughts?

**Comment by smalluban** [comment]
Hey!

Unfortunately, I wasn't aware of that flag. I was pretty sure, that the only one is the `"mode"`. I've been thinking for some time about extending `render` factory, but about providing a possibility to opt-out from shadowRoot at all, and allow to render into the inner light DOM content of the element. Personally, I don't see use cases for that, but I saw some request on other projects, which creates web components.

My idea is simple - add the second argument to `render` factory, which would be options object. Then using it directly (not by the translation) would support passing options:

```javascript
import { render, html } from 'hybrids';

export default {
   render: render(
      ({ value }) => html`<div>${value}</div>`,
      { delegatesFocus: true }, // The question is how it should look like?
   ),
}
```
You can use `render` factory directly with current version of the library, but it has no sence, as it is the same as setting `render` property with just a function.

As you are one of the users of the lib, what do you think about it? I think about two solutions for options object shape.

1. Every option is set explicit in code and delegate to `attachShadow` one by one (if in the future will be more of them). In that case, also `useShadow` should be added as well, with default to `true`.

2. Create `shadowRoot` option, which is an object or boolean. If `false` it not uses shadow DOM at all, if true it is as it now (and also default, for no breaking changes), and if object, its content is passed to `attachShadow` method like this:

```javascript
host.attachShadow({ mode: 'open', ...options });
```

**Comment by amarajs** [comment]
To your first point, we _have_ been writing a few "abstract" (i.e. behavior-only components without any visual rendering) that wrap normal light DOM. It's a great feature, and we get it using Hybrids by simply _not_ providing a render property:

```javascript
const toggleActive = (e) => {
    e.stopPropagation();
    e.currentTarget.active = !e.currentTarget.active;
};
const addToggleHandler = (host) => {
    host.addEventListener('click', toggleActive);
    return () => host.removeEventListener('click', toggleActive);
};
define('my-active-component', {
    active: property(false, addToggleHandler)
});
```
```html
  <my-active-component active>
      <!-- we can style buttons inside an active ancestor as if they were "pressed" -->
      <button>hello</button>
  </my-active-component>
```
So definitely keep this feature!

To your second point, since we can skip shadowRoot today by not specifying a render property, I'm less interested in a `useShadow` property. Instead, if you are already thinking of adding a new top-level property to the Hybrids definition object then let's make sure it supports additional possible extension points in the future (not just those related to creating a shadow root):

```javascript
define('my-component', {
    options: { shadowRoot: { delegatesFocus: true } },
    render: () => html`
        <dialog open>
            <button autofocus>this will be focused</button>
        </dialog>
    `
});
```
While we're on the subject, how would you feel about letting the render method optionally produce light DOM instead of shadow DOM? For example:

```javascript
define('my-component', {
    options: { shadowRoot: false },
    render: () => html`
        <p>This would be innerHTML, not shadow DOM.</p>
    `
});
```
The reason I'm asking is because the web components best practices that say not to put too much HTML into shadow DOM, but that makes writing larger "smart" components more difficult, since most smart components typically have a bunch of child components. It would be nice to output all those child elements as normal light DOM instead of shadow DOM, but with all that great lit-html functionality.

**Comment by amarajs** [comment]
Now that I look at it, it's probably not possible to assume an "options" property isn't already being used by some consumers. Maybe that information could be passed as the 3rd argument to `define`:

```javascript
const options = { shadowRoot: { delegatesFocus: true } };

define('my-component', {
    render: () => html`
        <dialog open>
            <button autofocus>this will be focused</button>
        </dialog>
    `
}, options);
```

**Comment by smalluban** [comment]
Render is just yet another **property** feature, so by adding options I ment extending `render` factory, not changing define method or adding new translated property. 

`options` should be a second argument of render factory. However, translation feature can't be used to set options, as render key is then just a function call. In that case direct use of render factory is required (as in my code example). 

My question was rather how shape that options object, but now I think the best option is the easiest and future ready, so it should be just 'shadowRoot' key with abality to pass bolean or object.

For your case usage would be like this:

```javascript
import { render, html } from 'hybrids';

export default {
  ...,
  render: render(() => html`...`, { shadowRoot: { delegatesFocus: true } }),
}
```

**Comment by daninnewyork** [comment]
Looks good! If someone passes { shadowRoot: false } to `render`, would it output innerHTML? That would be a great feature.

**Comment by smalluban** [comment]
I have bad news. ShadyDOM does not support `delegatesFocus` option. The only option that it uses is `"mode"` - look at here: https://github.com/webcomponents/shadydom/blob/e0e5db6cdd23aed5df30b49d73f309b179b7e4bd/src/attach-shadow.js#L58.

I've started working on this, but focus test fails for polyfilled browsers. I am not sure about introducing a feature, which is not usable now for all supported browsers.

**Comment by amarajs** [comment]
Thanks for opening that issue with the polyfill authors! Like you, I'm surprised it wasn't implemented already.

In any case, I don't think this feature should depend on the polyfill matching the spec. That's because you're not really implementing delegatesFocus -- your feature is to pass along the user's shadowRoot options to the attachShadow call. The unit tests for that shouldn't depend on the polyfill code.

What if a 3rd option were added to attachShadow later? You wouldn't write a new unit test to make sure the polyfill implemented it correctly -- that's their job! There may even be options in the future that a polyfill couldn't support -- should Hybrids not support those, either?

I think Hybrids should just pass along the user's options and let the consumer decide if they want to use a feature that isn't yet supported by the polyfill. Depending on the browsers they need to support, the polyfill may not even be a requirement.

In other words, I agree that all Hybrids features should be supported in polyfilled browsers, but I don't think this feature is "implement delegatesFocus" -- I think this feature is "pass { shadowRoot } options to the attachShadow method," which should have no problem passing tests in polyfilled browsers.

**Comment by smalluban** [comment]
Thank you for your comment. You are right, maybe I was to much focus on that particular feature. I will make a test for passing some options, whatever they are. If you don't mind, I will create a PR and add you as a reviewer?
